### ISSUE TITLE
Consider an approach different than socat for non-root user access docker

### ISSUE THREAD

**Issue by felipecassiors** [microsoft/vscode-dev-containers#488]
I'm talking about this: https://github.com/microsoft/vscode-dev-containers/tree/master/containers/docker-from-docker#enabling-non-root-access-to-docker-in-the-container.

The `socat` approach works fine, but there is also another one that works well. It consists of adjusting the `docker` group from the container to match the `docker` group from the user.

It's kinda tricky, but I made a small project to prove that it works. See https://github.com/felipecassiors/fixdockergid.

At the end, you have a very less intrusive `docker run` call:
```sh-session
docker run -u "$(id -u):$(id -g)" -v /var/run/docker.sock:/var/run/docker.sock ubuntu-with-fixdockergid
```

And the Dockerfile can still have the `USER vscode` in its end.

**Comment by PavelSosin-320** [comment]
I vote for it with another use-case: "Native docker running inside WSL2 distro (CentOS) listening standard /var/run/docker.sock:/var/run/docker.sock."  Docker itself is much more stable and secure than Docker-for-Windows due to pure enterprise origin and docker CLI and docker-compose come as platform packages.

**Comment by felipecassiors** [comment]
> I vote for it with another use-case: "Native docker running inside WSL2 distro (CentOS) listening standard /var/run/docker.sock:/var/run/docker.sock." Docker itself is much more stable and secure than Docker-for-Windows due to pure enterprise origin and docker CLI and docker-compose come as platform packages.

I didn't quite understand... can you elaborate?

**Comment by Chuxel** [comment]
@felipecassiors The script that is referenced is a bit smarter. If it detects the group is not "0", it adds the user to it. I don't want to just chown the socket since on Linux that would update the host. The socat version is a final fallback since I'm finding that on Windows and macOS, the group is zero.  That isn't really needed on Linux in most cases.

Check out https://github.com/microsoft/vscode-dev-containers/blob/master/script-library/docker-debian.sh

**Comment by felipecassiors** [comment]
@Chuxel you're right. The script already performs the necessary changes to the `docker` group in the container or even creates it.

The disadvantage of it against the `fixdockergid` approach is that it relies on either:

1. The container being started as `root`
2. `sudo`

You can get rid of those requirements by using the `suid`. The `fixdockergid` is one example of how it can be used on shell scripts, with the `shc` compiler.

**Comment by Chuxel** [comment]
Yeah, the the issue is extensions like "Docker" access the socket directly rather than through the CLI, so I'm not sure how setting that bit resolves the problem. The specific problem here is that the socket itself.. and thus the underlying GID and UID on the host, is "root" -- intentionally.

That said, we could probably drop the use of sudo if we set the `suid` bit on the shell scripts - that's a good idea.

There's code that updates the UID/GID if the remoteUser is set in the core of the extension, but this doesn't try to scan the whole filesystem to make updates given the potential perf implications of that. Not sure if @chrmarti looked at using `fixuid` for that functionality already.

**Comment by felipecassiors** [comment]
> That said, we could probably drop the use of sudo if we set the `suid` bit on the shell scripts - that's a good idea.

Exactly, that's the point.

> There's code that updates the UID/GID if the remoteUser is set in the core of the extension, but this doesn't try to scan the whole filesystem to make updates given the potential perf implications of that. Not sure if @chrmarti looked at using `fixuid` for that functionality already.

There is no much need to bother: [`fixuid`](https://github.com/boxboat/fixuid) does that all for you (updates the container's user to match the host's user, adjusts the file permissions, sets some envs). The `fixdockergid` mentioned is just an extension to it [until the feature gets merged into it](https://github.com/boxboat/fixuid/issues/29).

**Comment by felipecassiors** [comment]
I rewrote this Issue to point the real proposal.

**Comment by PavelSosin-320** [comment]
The standard way to enable access to the docker by any user is documented in every "native" docker installation guide: docker installation creates a docker user group and every user who belongs to this group has access to docker. For example, I run native docker in CentOS 7 distro that has my user, created and assigned to the docker user group and I can use docker without any sudo or wsl -u root just because my login user belongs to the docker group. The assignment is done once because users and groups are persistent in the Linux distro.
Indeed, I disabled SELINUX in my distro because I don't see such a security level is relevant for WSL. It means some security compromise.
When docker or podman are installed on SELINUX distro as a platform package from the corresponding repository the installation takes MAC security into account and such docker installation can pass the most rigorous security check.

**Comment by felipecrs** [comment]
Another problem which would be solved by using `fixuid` or at least it's approach is that we would not need to know the UID and GID of the host machine in the build time, but only in the run time.

This would enable people to build and push their non-root dev containers and be able to use anywhere else. I'm personally facing a problem trying to run some automated tests against my `dotfiles` using a pre-built Remote - Containers image (https://github.com/felipecrs/dotfiles/pull/16).

Because the user of GitHub Actions Runner has the UID 1001 and GID 116, I'm not able to start the container as a non-root user since the UID (`1000`) and GID (`1000`) of the `vscode` user in the container does not get properly adjusted at the run time.
